include/aspect/particle/world.h

```diff
@@ -21,729 +21,310 @@
-              old_types[i] = MPI_DOUBLE;
-            }
-
-          // Create and commit the MPI type
-          res = MPI_Type_struct(num_entries, block_lens, indices, old_types, &particle_type);
-          if (res != MPI_SUCCESS) exit(-1);
-
-          res = MPI_Type_commit(&particle_type);
-          if (res != MPI_SUCCESS) exit(-1);
-
-          // Delete temporary arrays
-          delete [] old_types;
-          delete [] indices;
-          delete [] block_lens;
-
-          // Initialize send/recv structures appropriately
-          num_send = new int[world_size];
-          num_recv = new int[world_size];
-          send_offset = new int[world_size];
-          recv_offset = new int[world_size];
-          send_reqs = new MPI_Request[world_size];
-          recv_reqs = new MPI_Request[world_size];
-        };
+        void
+        load_tracers(const typename parallel::distributed::Triangulation<dim>::cell_iterator &cell,
+                     const typename parallel::distributed::Triangulation<dim>::CellStatus status,
+                     const void *data);
 
         /**
-         * Calculate the cells containing each particle for all particles.
+         * Serialize the contents of this class.
          */
-        void find_all_cells()
-        {
-          typename std::multimap<LevelInd, T>::iterator   it;
-          std::multimap<LevelInd, T>                      tmp_map;
-          LevelInd                                        found_cell;
-
-          // Find the cells that the particles moved to
-          tmp_map.clear();
-          for (it=particles.begin(); it!=particles.end();)
-            {
-              found_cell = find_cell(it->second, it->first);
-              if (found_cell != it->first)
-                {
-                  tmp_map.insert(std::make_pair(found_cell, it->second));
-                  particles.erase(it++);
-                }
-              else ++it;
-            }
-          particles.insert(tmp_map.begin(),tmp_map.end());
-        };
+        template <class Archive>
+        void serialize (Archive &ar, const unsigned int version);
 
+      private:
         /**
-         * Advance particles by the specified timestep using the current
-         * integration scheme.
-         *
-         * @param [in] timestep Length of timestep to integrate particle
-         * movement
-         * @param [in] solution Current Aspect solution vector
+         * Integration scheme for moving particles in this world
          */
-        void advance_timestep(const double timestep, const LinearAlgebra::BlockVector &solution)
-        {
-          bool        continue_integrator = true;
-
-          // Find the cells that the particles moved to
-          find_all_cells();
-
-          // If the triangulation changed, we may need to move particles between processors
-          if (triangulation_changed) send_recv_particles();
+        Integrator::Interface<dim>     *integrator;
```

Comment:
Where do you get this (and the next few variables) from? I'd be more comfortable if they had a smart pointer type. If you create these variables yourself, use std_cxx11::unique_ptr, otherwise std_cxx11::shared_ptr.